STMReco/src/STMMovingWindowDeconvolution_module.cc

```diff
@@ -0,0 +1,293 @@
+    double _M; // M-parameter (used in differentiation step)
+    double _L; // L-parameter (used in averaging step)
+    double _nsigma_cut; // number of sigma away from baseline mean to cut (used in find_peaks)
+    double _thresholdgrad; // threshold on gradient
+
+    std::string _xAxis; // optional parameter for x-axis unit if plotting histograms
+  };
+
+  STMMovingWindowDeconvolution::STMMovingWindowDeconvolution(const Parameters& config ) :
+    art::EDProducer{config}
+    ,_verbosityLevel(config().verbosityLevel())
+    ,_stmWaveformDigisTag(config().stmWaveformDigisTag())
+    ,_tau(config().tau())
+    ,_M(config().M())
+    ,_L(config().L())
+    ,_nsigma_cut(config().nsigma_cut())
+    ,_thresholdgrad(config().thresholdgrad())
+  {
+    consumes<STMWaveformDigiCollection>(_stmWaveformDigisTag);
+    produces<STMMWDDigiCollection>();
+
+    _channel = STMUtils::getChannel(_stmWaveformDigisTag);
+
+    if (!config().xAxis(_xAxis)) {
+      if (_verbosityLevel >= 5) {
+        throw cet::exception("STMMovingWindowDecomposition") << "No xAxis scale defined despite requesting verbosity level >= 5" << std::endl;
+      }
+    }
+  }
+
+  void STMMovingWindowDeconvolution::beginJob() {
+  }
+
+  void STMMovingWindowDeconvolution::produce(art::Event& event) {
+    // create output
+    unique_ptr<STMMWDDigiCollection> outputMWDDigis(new STMMWDDigiCollection);
+    auto waveformDigisHandle = event.getValidHandle<STMWaveformDigiCollection>(_stmWaveformDigisTag);
+
+    STMEnergyCalib const& stmEnergyCalib = _stmEnergyCalib_h.get(event.id()); // get prodition
+
+    std::vector<double> deconvolved_data;
+    std::vector<double> differentiated_data;
+    std::vector<double> averaged_data;
+    int count = 0;
+    for (const auto& waveform : *waveformDigisHandle) {
+
+      // clear out data from previous waveform
+      deconvolved_data.clear();
+      deconvolved_data.reserve(waveform.adcs().size());
+      differentiated_data.clear();
+      differentiated_data.reserve(waveform.adcs().size());
+      averaged_data.clear();
+      averaged_data.reserve(waveform.adcs().size());
+
+      deconvolve(waveform, deconvolved_data, stmEnergyCalib);
+      differentiate(deconvolved_data, differentiated_data);
+      average(differentiated_data, averaged_data);
+
+      double baseline_mean = 0;
+      double baseline_stddev = 0;
+      calculate_baseline(averaged_data, baseline_mean, baseline_stddev);
+
+      std::vector<double> peak_heights;
+      std::vector<double> peak_times;
+      find_peaks(averaged_data, peak_heights, peak_times, baseline_mean, baseline_stddev);
+      for (size_t i_peak = 0; i_peak < peak_heights.size(); ++i_peak) {
+        STMMWDDigi mwd_digi(peak_times[i_peak], -1*peak_heights[i_peak]); // peak_heights are negative, make them positive here
+        outputMWDDigis->push_back(mwd_digi);
+      }
+
+      if (_verbosityLevel >= 5) {
+        art::ServiceHandle<art::TFileService> tfs;
```

Comment:
Are these histograms documenting transient state, which case they must be in this module and not in a companion module that works from data products?

If they do need to be in this module, I think it's good practice to move them to their own function so that the mainline reads cleanly.  This might mean a lot of arguments but that's OK - you only pay the runtime price of setting up the arguments when you are actually making histograms,

Reply:
These are documenting each step of the algorithm so we can see what it's doing. I have moved the histogram creation to a separate function in commit [b8917b5](https://github.com/Mu2e/Offline/pull/938/commits/b8917b511ee69cff93abc036855b717b3476d63d)